x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java

```diff
@@ -103,16 +215,239 @@ public AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting
     }
 
     @Override
-    public void cleanCaches() {}
+    public void cleanCaches() {
+        asyncFetchStore.clear();
+    }
 
     @Override
-    public void applyStartedShards(List<ShardRouting> startedShards, RoutingAllocation allocation) {}
+    public void applyStartedShards(List<ShardRouting> startedShards, RoutingAllocation allocation) {
+        for (ShardRouting startedShard : startedShards) {
+            asyncFetchStore.remove(startedShard.shardId());
+        }
+    }
 
     @Override
-    public void applyFailedShards(List<FailedShard> failedShards, RoutingAllocation allocation) {}
+    public void applyFailedShards(List<FailedShard> failedShards, RoutingAllocation allocation) {
+        for (FailedShard failedShard : failedShards) {
+            asyncFetchStore.remove(failedShard.getRoutingEntry().shardId());
+        }
+    }
 
     @Override
     public int getNumberOfInFlightFetches() {
-        return 0;
+        int count = 0;
+        for (AsyncCacheStatusFetch fetch : asyncFetchStore.values()) {
+            count += fetch.numberOfInFlightFetches();
+        }
+        return count;
+    }
+
+    private AsyncShardFetch.FetchResult<NodeCacheFilesMetadata> fetchData(ShardRouting shard, RoutingAllocation allocation) {
+        final ShardId shardId = shard.shardId();
+        final Settings indexSettings = allocation.metadata().index(shard.index()).getSettings();
+        final SnapshotId snapshotId = new SnapshotId(
```

Comment:
should we `assert SearchableSnapshotsConstants.isSearchableSnapshotStore(indexSettings)`, just in case?